Source/Core/VideoCommon/TextureCacheBase.cpp

```diff
@@ -1032,61 +1032,106 @@ void TextureCacheBase::CopyRenderTargetToTexture(u32 dstAddr, unsigned int dstFo
 
-	entry->frameCount = FRAMECOUNT_INVALID;
-	entry->SetEfbCopy(dstStride);
-	entry->is_custom_tex = false;
+	// RGBA takes two cache lines per block; all others take one
+	const u32 bytes_per_block = dstFormat == GX_TF_RGBA8 ? 64 : 32;
 
-	entry->FromRenderTarget(dst, dstFormat, dstStride, srcFormat, srcRect, isIntensity, scaleByHalf, cbufid, colmat);
+	u32 bytes_per_row = num_blocks_x * bytes_per_block;
 
-	u64 hash = entry->CalculateHash();
-	entry->SetHashes(hash, hash);
+	bool copy_to_ram = !g_ActiveConfig.bSkipEFBCopyToRam;
+	bool copy_to_vram = true;
+
+	if (copy_to_ram)
+	{
+		g_texture_cache->CopyEFB(
+			dst,
+			dstFormat,
+			tex_w,
+			bytes_per_row,
+			num_blocks_y,
+			dstStride,
+			srcFormat,
+			srcRect,
+			isIntensity,
+			scaleByHalf);
+	}
+	else
+	{
+		u8* ptr = dst;
+		for (u32 i = 0; i < num_blocks_y; i++)
+		{
+			memset(ptr, 0, bytes_per_row);
+			ptr += dstStride;
+		}
+	}
+
+	if (g_bRecordFifoData)
+	{
+		// Mark the memory behind this efb copy as dynamicly generated for the Fifo log
+		u32 address = dstAddr;
+		for (u32 i = 0; i < num_blocks_y; i++)
+		{
+			FifoRecorder::GetInstance().UseMemory(address, bytes_per_row, MemoryUpdate::TEXTURE_MAP, true);
+			address += dstStride;
+		}
+	}
 
 	// Invalidate all textures that overlap the range of our efb copy.
 	// Unless our efb copy has a weird stride, then we want avoid invalidating textures which
 	// we might be able to do a partial texture update on.
-	if (entry->memory_stride == entry->BytesPerRow())
+	if (dstStride == bytes_per_row || !copy_to_vram)
 	{
 		TexCache::iterator iter = textures_by_address.begin();
 		while (iter != textures_by_address.end())
 		{
-			if (iter->second->OverlapsMemoryRange(dstAddr, entry->size_in_bytes))
-				iter = FreeTexture(iter);
-			else
+			if (iter->second->addr + iter->second->size_in_bytes <= dstAddr || iter->second->addr >= dstAddr /* + size_in_bytes*/ )
 				++iter;
+			else
+				iter = FreeTexture(iter);
 		}
 	}
 
-	if (g_ActiveConfig.bDumpEFBTarget)
+	if (copy_to_vram)
 	{
-		static int count = 0;
-		entry->Save(StringFromFormat("%sefb_frame_%i.png", File::GetUserPath(D_DUMPTEXTURES_IDX).c_str(),
-			count++), 0);
-	}
+		// create the texture
+		TCacheEntryConfig config;
+		config.rendertarget = true;
+		config.width = scaled_tex_w;
+		config.height = scaled_tex_h;
+		config.layers = FramebufferManagerBase::GetEFBLayers();
 
-	if (g_bRecordFifoData)
-	{
-		// Mark the memory behind this efb copy as dynamicly generated for the Fifo log
-		u32 address = dstAddr;
-		for (u32 i = 0; i < entry->NumBlocksY(); i++)
+		TCacheEntryBase* entry = AllocateTexture(config);
+
+		entry->SetGeneralParameters(dstAddr, 0, dstFormat);
+		entry->SetDimensions(tex_w, tex_h, 1);
+
+		entry->frameCount = FRAMECOUNT_INVALID;
+		entry->SetEfbCopy(dstStride);
+		entry->is_custom_tex = false;
+
+		entry->FromRenderTarget(dst, dstFormat, dstStride, srcFormat, srcRect, isIntensity, scaleByHalf, cbufid, colmat);
```

Comment:
This function doesn't need: dstFormat and isIntensity as they are encoded in cbufid/colmat

It also doesn't need dstStride, as it's not dealing with actual ram.